DataFormats/Detectors/GlobalTracking/src/RecoContainer.cxx

```diff
@@ -513,16 +543,7 @@ void RecoContainer::fillTrackMCLabels(const gsl::span<GTrackID> gids, std::vecto
   }
 }
 
-void o2::globaltracking::RecoContainer::createTracks(std::function<bool(const o2::track::TrackParCov&, o2::dataformats::GlobalTrackID)> const& creator) const
```

Comment:
I would actually keep at least the createTracks. It is convenient, and does not require additional header and code duplication. I think the other one can go.
You could rename it to createBarrelTracks and call the callback in the lambda only with the same if consteppr check for TrackParCov as you did for the GPUHelper.

Reply:
Hi @davidrohr , createTracks is specialized to `o2::track::TrackParCov`, which has a problem with the createTracksVariadic now dealing also with `o2::track::TrackParCovFwd`, as we discussed by e-mail. To keep such specialization one needs to isolate the MFT part of `createTracksVariadic`, which is not obvious.

Reply:
@rpezzi : I think with this patch it should work:
```
diff --git a/DataFormats/Detectors/GlobalTracking/src/RecoContainer.cxx b/DataFormats/Detectors/GlobalTracking/src/RecoContainer.cxx
index f1009dd9f33..4397c3d28d5 100644
--- a/DataFormats/Detectors/GlobalTracking/src/RecoContainer.cxx
+++ b/DataFormats/Detectors/GlobalTracking/src/RecoContainer.cxx
@@ -513,9 +513,15 @@ void RecoContainer::fillTrackMCLabels(const gsl::span<GTrackID> gids, std::vecto
   }
 }
 
-void o2::globaltracking::RecoContainer::createTracks(std::function<bool(const o2::track::TrackParCov&, o2::dataformats::GlobalTrackID)> const& creator) const
-{
-  createTracksVariadic([&creator](const o2::track::TrackParCov& _tr, GTrackID _origID, float t0, float terr) { return creator(_tr, _origID); });
+void o2::globaltracking::RecoContainer::createBarrelTracks(std::function<bool(const o2::track::TrackParCov&, o2::dataformats::GlobalTrackID)> const& creator) const
+{
+  createTracksVariadic([&creator](const o2::track::TrackParCov& _tr, GTrackID _origID, float t0, float terr) {
+    if constexpr (std::is_base_of_v<o2::track::TrackParCov, std::decay_t<decltype(_tr)>>) {
+      return creator(_tr, _origID);
+    } else {
+      return false;
+    }
+  });
 }
 
 void o2::globaltracking::RecoContainer::createTracksWithMatchingTimeInfo(std::function<bool(const o2::track::TrackParCov&, GTrackID, float, float)> const& creator) const

```

Reply:
@davidrohr , unfortunately this does not work, Note that `createTracksVariadic` remains unprotected by the `if constexpr`. The compiler wants to generate a `createTracksVariadic` specialized for `TrackParCov`.

Reply:
ok, sure, my fault, sorry, that was dumb... But this should work then:
```
diff --git a/DataFormats/Detectors/GlobalTracking/src/RecoContainer.cxx b/DataFormats/Detectors/GlobalTracking/src/RecoContainer.cxx
index f1009dd9f33..4397c3d28d5 100644
--- a/DataFormats/Detectors/GlobalTracking/src/RecoContainer.cxx
+++ b/DataFormats/Detectors/GlobalTracking/src/RecoContainer.cxx
@@ -513,9 +513,15 @@ void RecoContainer::fillTrackMCLabels(const gsl::span<GTrackID> gids, std::vecto
   }
 }
 
-void o2::globaltracking::RecoContainer::createTracks(std::function<bool(const o2::track::TrackParCov&, o2::dataformats::GlobalTrackID)> const& creator) const
-{
-  createTracksVariadic([&creator](const o2::track::TrackParCov& _tr, GTrackID _origID, float t0, float terr) { return creator(_tr, _origID); });
+void o2::globaltracking::RecoContainer::createBarrelTracks(std::function<bool(const o2::track::TrackParCov&, o2::dataformats::GlobalTrackID)> const& creator) const
+{
+  createTracksVariadic([&creator](const auto& _tr, GTrackID _origID, float t0, float terr) {
+    if constexpr (std::is_base_of_v<o2::track::TrackParCov, std::decay_t<decltype(_tr)>>) {
+      return creator(_tr, _origID);
+    } else {
+      return false;
+    }
+  });
 }
 
 void o2::globaltracking::RecoContainer::createTracksWithMatchingTimeInfo(std::function<bool(const o2::track::TrackParCov&, GTrackID, float, float)> const& creator) const
```

Reply:
Thanks. This works! Commit added.